brain/synapse.h

```diff
@@ -0,0 +1,138 @@
+     *         index in the afferent contacts array.
+     * @throw std::runtime_error if index information not found in the synapse
+     *                           source of the circuit.
+     */
+    BRAIN_API SynapseGID getGID() const;
+
+    /** @name Presynaptic topological and geometrical attributes */
+    //@{
+    /** @return the GID of the presynaptic neuron. */
+    BRAIN_API uint32_t getPresynapticGID() const;
+
+    /** @return the section ID on the presynaptic neuron. */
+    BRAIN_API uint32_t getPresynapticSectionID() const;
+
+    /** @return the segment ID on the presynaptic neuron. */
+    BRAIN_API uint32_t getPresynapticSegmentID() const;
+
+    /** @return the distance in micrometer to the presynaptic neuron. */
+    BRAIN_API float getPresynapticDistance() const;
+
+    /** @return the presynaptic touch position on the surface of the segment. */
+    BRAIN_API Vector3f getPresynapticSurfacePosition() const;
+
+    /** @return the presynaptic touch position in the center of the segment. */
+    BRAIN_API Vector3f getPresynapticCenterPosition() const;
+    //@}
+
+    /** @name Presynaptic topological and geometrical attributes */
+    //@{
+    /** @return the GID of the postsynaptic neuron. */
+    BRAIN_API uint32_t getPostsynapticGID() const;
+
+    /** @return the section ID on the postsynaptic neuron. */
+    BRAIN_API uint32_t getPostsynapticSectionID() const;
+
+    /** @return the segment ID on the postsynaptic neuron. */
+    BRAIN_API uint32_t getPostsynapticSegmentID() const;
+
+    /** @return the distance in micrometer to the postsynaptic neuron. */
+    BRAIN_API float getPostsynapticDistance() const;
+
+    /** @return the postsynaptic touch position on the surface of the segment.*/
+    BRAIN_API Vector3f getPostsynapticSurfacePosition() const;
+
+    /** @return the postsynaptic touch position in the center of the segment. */
+    BRAIN_API Vector3f getPostsynapticCenterPosition() const;
+    //@}
+
+    /** @name Synaptic model attributes */
+    //@{
+    /** @return the axonal delay in milliseconds. */
+    BRAIN_API float getDelay() const;
+
+    /** @return the conductance in nanosiemens. */
+    BRAIN_API float getConductance() const;
+
+    /** @return the neuro-transmitter release probability. */
+    BRAIN_API float getUtilization() const;
+
+    /** @return the depression time constant in milliseconds. */
+    BRAIN_API float getDepression() const;
+
+    /** @return the facilitation time constant in milliseconds. */
+    BRAIN_API float getFacilitation() const;
+
+    /** @return the decay time constant in milliseconds. */
+    BRAIN_API float getDecay() const;
+
+    /** @return the absolute synaptic efficacy in millivolts. */
+    BRAIN_API int getEfficacy() const;
+    //@}
+
+private:
+    Synapse& operator=( const Synapse& ) = delete;
+
+    friend class Synapses;
+    Synapse( const Synapses& synapses, size_t index );
+
+    struct Impl;
+    std::unique_ptr< const Impl > _impl;
```

Comment:
I'm hesitating whether you should use pimpl here. I don't expect many changes in the implementation of this object and for performance reasons it's better to not use the heap and simply store the Synapses& and index here. On the other hand, if performance is a concern, you should use the array interface in Synapses instead.